msdk-io/msdk-io-netcdf/src/main/java/io/github/msdk/io/netcdf/NetCDFMsScan.java

```diff
@@ -54,51 +54,70 @@ public NetCDFMsScan(Integer scanNumber, int[] scanStartPositions, float[] scanRe
 
   @Override
   public float[] getIntensityValues(float[] intensityValues) {
-    final Integer scanIndex = getScanIndex();
-    numOfDataPoints = getNumberOfDataPoints();
-    try {
-      final int scanStartPosition[] = {scanStartPositions[scanIndex]};
-      final int scanLength[] = {scanStartPositions[scanIndex + 1] - scanStartPositions[scanIndex]};
-      final Array intensityValueArray = intensityValueVariable.read(scanStartPosition, scanLength);
-      final Index intensityValuesIndex = intensityValueArray.getIndex();
-
+    if (this.intensityValues == null) {
+      final Integer scanIndex = getScanIndex();
+      numOfDataPoints = getNumberOfDataPoints();
+      try {
+        final int scanStartPosition[] = {scanStartPositions[scanIndex]};
+        final int scanLength[] =
+            {scanStartPositions[scanIndex + 1] - scanStartPositions[scanIndex]};
+        final Array intensityValueArray =
+            intensityValueVariable.read(scanStartPosition, scanLength);
+        final Index intensityValuesIndex = intensityValueArray.getIndex();
+
+        if (intensityValues == null || intensityValues.length < numOfDataPoints)
+          intensityValues = new float[numOfDataPoints];
+
+        // Load the data points
+        for (int i = 0; i < numOfDataPoints; i++) {
+          final Index intensityIndex0 = intensityValuesIndex.set0(i);
+          intensityValues[i] =
+              (float) (intensityValueArray.getDouble(intensityIndex0) * intensityValueScaleFactor);
+        }
+      } catch (IOException | InvalidRangeException e) {
+        throw new MSDKRuntimeException(e);
+      }
+    } else {
       if (intensityValues == null || intensityValues.length < numOfDataPoints)
```

Comment:
The `getIntensityValues()` function calls this function as `getIntensityValues(null)`. So, it is necessary to check if the passed array is null before checking if the data fits into it.